staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,388 @@
+	}
+	c.setStream(id, s)
+	return s, nil
+}
+
+// sendPings starts heartbeat sending a "ping" from the endpoint every "period".
+func (c *wsStreamCreator) sendPings(period time.Duration) {
+	c.connMu.Lock()
+	closedCh := make(chan struct{})
+	closeHandler := c.conn.CloseHandler()
+	c.conn.SetCloseHandler(func(code int, text string) error {
+		close(closedCh)
+		return closeHandler(code, text)
+	})
+	c.connMu.Unlock()
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-closedCh:
+			return
+		case <-t.C:
+			c.connMu.Lock()
+			if err := c.conn.WriteControl(gwebsocket.PingMessage, nil, time.Now().Add(period)); err != nil {
+				klog.V(7).Infof("Websocket Ping failed: %v", err)
+				// Continue, in case this is a transient failure.
+				// c.conn.CloseChan above will tell us when the connection is
+				// actually closed.
+			} else {
+				klog.V(7).Infof("Websocket Ping succeeeded")
+			}
+			c.connMu.Unlock()
+		}
+	}
+}
+
+// run is executed in its own goroutine, reading the connection and demultiplexing
+// the data messages into individual streams.
+func (c *wsStreamCreator) run(bufferSize int) {
+	// Buffer size must correspond to the same size allocated
+	// for the read buffer during websocket client creation. A
+	// difference can cause incomplete connection reads.
+	readBuffer := make([]byte, bufferSize)
+	// Set up handler for ping/pong heartbeat.
+	if err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline)); err != nil {
+		klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+	}
+	c.conn.SetPongHandler(func(string) error {
+		err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline))
+		if err != nil {
+			klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+		}
+		return nil
+	})
+	for {
+		messageType, r, err := c.conn.NextReader()
+		if err != nil {
+			websocketErr, ok := err.(*gwebsocket.CloseError)
+			if ok && websocketErr.Code == gwebsocket.CloseNormalClosure {
+				err = nil // readers will get io.EOF as it's a normal closure
+			} else {
+				err = fmt.Errorf("next reader: %w", err)
+			}
+			c.closeAllStreamReaders(err)
+			return
+		}
+		if messageType != gwebsocket.BinaryMessage {
+			c.closeAllStreamReaders(fmt.Errorf("unexpected message type: %d", messageType))
+			return
+		}
+		// it's ok to read just a single byte because the underlying library wraps the actual connection with
+		// a buffered reader anyway.
+		_, err = io.ReadFull(r, readBuffer[:1])
+		if err != nil {
+			c.closeAllStreamReaders(fmt.Errorf("read stream id: %w", err))
+			return
+		}
+		streamID := readBuffer[0]
+		s := c.getStream(streamID)
+		if s == nil {
+			klog.Errorf("Unknown stream id %d, discarding message", streamID)
+			continue
+		}
+		for {
+			nr, errRead := r.Read(readBuffer)
+			if nr > 0 {
+				_, errWrite := s.writePipe.Write(readBuffer[:nr])
+				if errWrite != nil {
+					// Pipe must have been closed by the stream user. Nothing to do, discard the message.
+					break
```

Comment:
From the pipe write documentation:
```
// it writes data to the pipe, blocking until one or more readers                                                                                                    
// have consumed all the data or the read end is closed.                                                                                                             
// If the read end is closed with an error, that err is                                                                                                              
// returned as err; otherwise err is ErrClosedPipe.           
```
According to this, the error is *always* that the read end is closed. The type of close on the read end determines the error.